packages/components/src/composite/legacy.tsx

```diff
@@ -0,0 +1,301 @@
+	setShift: ( v: CompositeState[ 'shift' ] ) => void;
+	wrap: boolean | Orientation;
+	setWrap: ( v: CompositeState[ 'wrap' ] ) => void;
+}
+
+type FilteredCompositeStateProps = Pick< CompositeState, 'store' | 'baseId' >;
+
+type PartialCompositeState = Partial<
+	Omit< CompositeState, keyof FilteredCompositeStateProps >
+> &
+	FilteredCompositeStateProps;
+
+export type CompositeStateProps =
+	| { state: PartialCompositeState }
+	| ( PartialCompositeState & { state?: never } );
+
+interface IdState {
+	baseId: string;
+	setBaseId: React.Dispatch< React.SetStateAction< string > >;
+}
+
+function showDeprecationMessage( ...props: Parameters< typeof deprecated > ) {
+	if ( 'test' !== process.env.NODE_ENV ) {
+		deprecated( ...props );
+	}
+}
+
+const idMap = new Map< string, React.MutableRefObject< number > >();
+
+function useBaseId(
+	initialState: Pick< Partial< IdState >, 'baseId' > = {}
+): IdState {
+	const { baseId: initialBaseId } = initialState;
+	const [ baseId, setBaseId ] = useState(
+		() =>
+			initialBaseId ||
+			// eslint-disable-next-line no-restricted-syntax
+			`id-${ Math.random().toString( 32 ).substring( 2, 8 ) }`
+	);
+	idMap.set( baseId, useRef( 0 ) );
+	return { baseId, setBaseId };
+}
+
+function useId( baseId: string, preferredId?: string ) {
+	const counter = idMap.get( baseId ) ?? { current: 0 };
+
+	const [ id, setId ] = useState( () => {
+		if ( preferredId ) return preferredId;
+		return `${ baseId }-${ ++counter.current }`;
+	} );
+
+	return { id, setId };
+}
+
+function useMapInitialStateToStoreProps(
+	initialState: InitialState = {}
+): CompositeStoreProps & IdState {
+	const {
+		currentId: defaultActiveId,
+		orientation,
+		rtl = false,
+		loop: focusLoop = false,
+		wrap: focusWrap = false,
+		shift: focusShift = false,
+		// eslint-disable-next-line camelcase
+		unstable_virtual: virtualFocus,
+		...idState
+	} = initialState;
+
+	return {
+		...useBaseId( idState ),
+		defaultActiveId,
+		rtl,
+		orientation,
+		focusLoop,
+		focusShift,
+		focusWrap,
+		virtualFocus,
+
+		// TODO?
+		includesBaseElement: false,
+	};
+}
+
+function useCreateStoreFromInitialState(
+	initialState: InitialState = {}
+): FilteredCompositeStateProps & IdState {
+	const { baseId, setBaseId, ...storeProps } =
+		useMapInitialStateToStoreProps( initialState );
+	const store = Current.useCompositeStore( storeProps );
+	return { store, baseId, setBaseId };
+}
+
+function useStateBuilder<
+	Store extends CompositeStore,
+	Config extends Record< Key, Value >,
+	Key extends Parameters< Store[ 'setState' ] >[ 0 ],
+	Value extends [ string, string ],
+>( store: Store, config: Partial< Config > ) {
+	const meta = {} as Record< string, any >;
+	for ( const [ key, [ get, set ] ] of Object.entries< Value >(
+		config as Config
+	) ) {
+		meta[ get ] = store.useState( key as Key );
+		meta[ set ] = (
+			value: Parameters< typeof store.setState< Key > >[ 1 ]
+		) => {
+			store.setState< Key >( key as Key, value );
+		};
+	}
+	return meta;
+}
```

Comment:
This function is somewhat difficult to comprehend initially. Perhaps some comments could help?